scipy/sparse/linalg/_special_sparse_arrays.py

```diff
@@ -0,0 +1,394 @@
+import numpy as np
+from scipy.sparse.linalg import LinearOperator
+from scipy.sparse import spdiags, kron, eye, dia_array
+
+
+__all__ = ['LaplacianNd']
+
+
+class LaplacianNd(LinearOperator):
+    """
+    Construct a sparse matrix and a callable associated with the Laplacian
+    on a uniform rectangular grid in `N` dimensions given by tuple `shape`
+    and output its eigenvalues.
+
+    The Laplacian matrix `L` is square real symmetric negative definite.
+
+    Parameters
+    ----------
+    shape : tuple of int
+        A tuple of length `N` (corresponding to the dimension of the Lapacian),
+        where each entry gives the size of that dimension. The Laplacian matrix
+        is square of the size np.prod(shape)
+    boundary_conditions: ``'Dirichlet'`` or ``'Neumann'`` or ``'Periodic'``
+        The type of the boundary conditions on the boundaries of the grid
+    dtype: one of numeric numpy dtypes
+        The dtype of the matrix output.
+        Default: ``np.int32``.
+
+    Returns
+    -------
+    lap_Nd: `LinearOperator` custom object
+        The object containing the output standard for `LinearOperator`: e.g.,
+        lap_Nd.toarray() : (n, n) ndarray
+        lap_Nd.tosparse() : (n, n) DIAgonal sparse format
+        lap_Nd._matvec and lap_Nd._matmat: callable objects
+            The handles to functions that multiply the Laplacian matrix `L` of
+            the shape `n`-by-`n` on the right by an input matrix `x` of
+            the shape `n`-by-`k` to output ``L @ x`` without constructing `L`
+    lap_Nd.eigenvalues : (n, ) ndarray, float
+        Eigenvalues of the Laplacian matrix ordered ascending
+
+    laplacian_obj: class
+        The object containing the output
+    laplacian_obj.sparse : csr_array
+        The ``(n, n)`` Laplacian matrix in a Compressed Sparse Row (CSR) format
+    laplacian_obj : callable
+        The handle to a function that multiplies the Laplacian matrix
+        `L` of the shape `n`-by-`n` on the right by an input matrix `x`
+        of the shape `n`-by-`k` to output ``L @ x`` without constructing `L`
+    laplacian_obj.eigenvalues : (n, ) ndarray, float
+        Eigenvalues of the laplacian matrix (pair) ordered ascending
```

Comment:
I am not sure all of this is in the right place ie. Some are attributes so should be under an Attributes section and others belong in the docstring of the various methods

Reply:
@j-bowhay sorry this is still a mess with leftovers from the previous revision. I'll get it all cleaned up after cicrle-ci starts working and generates the documentation to look at.

Reply:
For reference, [here is our docstring standard for classes](https://numpydoc.readthedocs.io/en/latest/format.html#documenting-classes).

Reply:
I have done my best to follow the standard in the latest commit. If the format is still wrong, yes, I would of course appreciate your help. Please feel free to commit directly. 

There are a few changes requested in the examples, e.g., https://github.com/scipy/scipy/pull/18963#discussion_r1285299668 that I can do myself, but since you could test it locally I would appreciate if you edit this as well together with the proper formatting if needed. 

Thanks very much for fixing the refguide problem! So finally I can read the docs!